ApeWorX/ape
# feat: include view calls in gas report

### What I did



fixes: #1091 

### How I did it

1. For local geth dev, if the user is requesting trace-related info, use `debug_traceCall` in place of `send_call`
(The user can request this either via the `--gas` flag in `ape test` or via kwargs such as `show_trace` or `show_gas_report` while making a call.
3. Adds a protected (`_reports`) manager to `chain_manager`. This is NOT a public API (for now).
4. In `ReceiptCapture` (part of `pytest` plugin code), tell the report manager to track gas. Also, make the `gas_report` from `ReceiptCapture` to the `ReportManager`.
5. Add default implementation in `TestProviderAPI` that sends the call as a transaction, captures its gas, and then reverts it.

**WARN**: The `--gas` flag is really slow for Hardhat (the default implementation)... But using `geth`, it is really fast  I am not sure why that is... Hardhat, when using `--gas` and the changes in this PR, takes twice as long... I think the implementation to send all calls as transaction surrounded by snapshotting is just slow.. There are plans, however, to get `debug_traceCall` in Hardhat _and_ Foundry, so that will help a lot.

### How to verify it

* Use `--gas` flag in a project with Geth provider
* Use `--gas` flag in a project with Hardhat provider

You should now see a list full of the view methods called as well as the mutable ones.

### Checklist
